internal/transport/controlbuf.go

```diff
@@ -527,12 +530,16 @@ const minBatchSize = 1000
 // As an optimization, to increase the batch size for each flush, loopy yields the processor, once
 // if the batch size is too low to give stream goroutines a chance to fill it up.
 func (l *loopyWriter) run() (err error) {
-	// Always flush the writer before exiting in case there are pending frames
-	// to be sent.
-	defer l.framer.writer.Flush()
+	defer func() {
+		if logger.V(logLevel) {
+			logger.Infof("transport: loopyWriter exiting with error: %v", err)
+		}
+		l.cbuf.finish()
+	}()
 	for {
 		it, err := l.cbuf.get(true)
 		if err != nil {
+			l.closeConnection()
 			return err
 		}
 		if err = l.handle(it); err != nil {
```

Comment:
Also, I see that `processData()` today returns an error only when writing of data or headers fails. But how can we guarantee that in the future? Should we at least document  that `handle()` and `processData()` should return errors only for I/O related events. And also document `loopy.run()` saying it will close the connection only when it sees a non-I/O error.

Reply:
`handle` returns errors from the handlers themselves which has a lot of I/O error possibilities.  We'll need to do that closing in `handle` itself.  Or we could make the errors carry a type with a bool to indicate whether they are I/O errors, but that felt messier.

It seems there's two ways to do this.  Commit 1 is wrapping in a lot of different places which feels finnicky, and commit 2 is wrapping in the writer which might not work if the http2 framer decides to start wrapping errors without supporting `Unwrap` (but that seems very unlikely and we could deal with it if it ever happens).

Reply:
I prefer the `ioError` option as well, and the PR looks good to me.